### PR TITLE
[SPARK-45217][CORE] Support change log level of specific package or class

### PR DESCRIPTION
**Pull request by warrenzhu25**
### What changes were proposed in this pull request?
Add `SparkContext.setLogLevel(loggerName: String, logLevel: String)` to support change log level of specific package or class, it has below use cases:

1. Change log level in runtime without restart spark driver or executor. One example is spark-shell.
2. Save the effort to changing the log4j.properties on all host machines or passing long parameter in spark-submit to override log4j.properties.

### Why are the changes needed?
Currently, only support change log level of root logger, we may want to change log level of specific package or class

### Does this PR introduce _any_ user-facing change?
Yes, added below method in `SparkContext`
```
def setLogLevel(loggerName: String, logLevel: String): Unit
```

### How was this patch tested?
Added test in `UtilsSuite`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -1013,7 +1013,7 @@ private[spark] object Utils
    * In case of IPv6, getHostAddress may return '0:0:0:0:0:0:0:1'.
    */
   def localHostName(): String = {
-    addBracketsIfNeeded(customHostname.getOrElse(localIpAddress.getHostAddress))
+    addBracketsIfNeeded(customHostname.getOrElse(localIpAddress.getHostName))
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:1016]: This looks irrelevant to me. Why do you change this?

**Review by dongjoon-hyun — COMMENTED** [review]
Why don't we use `log4j2.properties`?

**Comment by warrenzhu25** [comment]
> Why don't we use `log4j2.properties`?

Added user cases in description.

**Comment by mridulm** [comment]
+CC @srowen

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -401,6 +402,23 @@ class SparkContext(config: SparkConf) extends Logging {
...
   }
 
+  /** Change logLevel of specific package or class name.
+   *  This overrides any user-defined log settings.
+   *
+   * @param loggerName package or class name such as "org.apache.spark" or
+   *                   "org.apache.spark.SparkContext"
+   * @param logLevel The desired log level as a string.
+   *                 Valid log levels include: ALL, DEBUG, ERROR, FATAL, INFO, OFF, TRACE, WARN
+   */
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala:412]: Need the "since 4.0.0" on the new methods

Comment (srowen) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:412]: We need this in Pyspark, no?

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/SparkContext.scala (hunk 2 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/SparkContext.scala:414]: Added since. I'll add this in Pyspark in followup PR.

**Comment by srowen** [comment]
Nah, let's address Python and R here

**Comment by warrenzhu25** [comment]
> Nah, let's address Python and R here

Added.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -2323,31 +2323,75 @@ private[spark] object Utils
...
   }
 
+  /**
+   * configure a new log4j level for specific package or class
+   *
+   * @param name package or class name such as "org.apache.spark" or
+   *             "org.apache.spark.SparkContext"
+   * @param level new level of [[org.apache.log4j.Level]]
+   */
+  def setLogLevel(name: String, level: Level): Unit = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:2341]: I don't think we need this override, do we? the original method didn't support this.
Id also imagine that the old method just just call this one with some value representing the root logger

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 4]
```diff
@@ -2323,31 +2323,75 @@ private[spark] object Utils
...
+    Option(loggerConfig).map(c => c.getLevel)
+  }
+
+  /**
+   * remove logger for specific package or class
+   *
+   * @param name package or class name such as "org.apache.spark" or
+   *             "org.apache.spark.SparkContext"
+   */
+  def removeLogger(name: String): Unit = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:2371]: Do you really need get/remove? not sure it's necessary

File: python/pyspark/context.py [hunk 5]
```diff
@@ -534,6 +534,27 @@ def setLogLevel(self, logLevel: str) -> None:
         """
         self._jsc.setLogLevel(logLevel)
 
+    def setLogLevel(self, logName: str, logLevel: str) -> None:
```
Comment (srowen) [inline python/pyspark/context.py:537]: Yeah, I mean, this is all you've exposed in R and Python, so not sure you should/need to expose more in Java/Scala

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 3 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2341]: Do you want me to remove old method `def setLogLevel(l: Level): Unit`, and calling new method instead?

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 4 above)
Comment (warrenzhu25) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2371]: Currently, get/remove in `Utils` only used by tests, but we might consider expose them in `SparkContext` in the future.

Comment (srowen) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2371]: They should be private to Spark then and just make sure they're annotated as for testing only

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 3 above)
Comment (srowen) [reply core/src/main/scala/org/apache/spark/util/Utils.scala:2341]: Oh, nevermind, there were already two overloads? then two new overloads makes sense.
But yeah they should call each other in a logical way

**Review by srowen — COMMENTED** [review]
The change looks about right, but there are some test failures, eg https://github.com/warrenzhu25/spark/actions/runs/6332926200/job/17202261106

**Comment by warrenzhu25** [comment]
> The change looks about right, but there are some test failures, eg https://github.com/warrenzhu25/spark/actions/runs/6332926200/job/17202261106

The example test should be skipped as I added `# doctest :+SKIP`. Any ideas why this happen?
